File: violin_coach_app/lib/utils/rhythm_trainer.dart
/// A utility class to handle rhythm training for Twinkle Twinkle Little Star pattern.
class RhythmTrainer {
  // Current progress through the song
  int _currentNoteIndex = 0;
  
  // Total notes in the song
  final int _totalNotes = 42;  // Twinkle Twinkle has 42 notes
  
  // Expected bow direction pattern for the song
  final List<String> _bowDirectionPattern = [
    'Down', 'Up', 'Down', 'Up', 'Down', 'Up', 'Down',  // First phrase
    'Up', 'Down', 'Up', 'Down', 'Up', 'Down', 'Up',    // Second phrase
    'Down', 'Up', 'Down', 'Up', 'Down', 'Up', 'Down',  // Third phrase
    'Up', 'Down', 'Up', 'Down', 'Up', 'Down', 'Up',    // Fourth phrase
    'Down', 'Up', 'Down', 'Up', 'Down', 'Up', 'Down',  // Fifth phrase
    'Up', 'Down', 'Up', 'Down', 'Up', 'Down', 'Up',    // Last phrase
  ];
  
  // Track correct and incorrect bow directions
  int _correctDirections = 0;
  int _incorrectDirections = 0;
  
  // Last direction we detected
  String _lastDirection = '';
  
  /// Update progress through the song based on detected bow direction
  Map<String, dynamic> updateProgress(String currentDirection) {
    // Don't count the same direction twice in a row
    if (currentDirection == _lastDirection) {
      return {
        'status': 'Waiting for new bow stroke...',
        'current_note': _currentNoteIndex + 1,
        'total_notes': _totalNotes,
        'progress_percent': (_currentNoteIndex / _totalNotes),
      };
    }
    
    // Store this direction
    _lastDirection = currentDirection;
    
    // Check if the direction is correct for the current note
    final expectedDirection = _bowDirectionPattern[_currentNoteIndex];
    final isCorrect = currentDirection == expectedDirection;
    
    // Update tracking
    if (isCorrect) {
      _correctDirections++;
    } else {
      _incorrectDirections++;
    }
    
    // Build status message
    String status;
    if (isCorrect) {
      status = 'Good! ${_getPositionInSong()}';
    } else {
      status = 'Try ${expectedDirection.toLowerCase()} bow for this note. ${_getPositionInSong()}';
    }
    
    // Move to next note regardless (to keep them progressing)
    _currentNoteIndex = (_currentNoteIndex + 1) % _totalNotes;
    
    // Calculate progress percentage
    final progressPercent = _currentNoteIndex / _totalNotes;
    
    // Return result
    return {
      'status': status,
      'is_correct': isCorrect,
      'current_note': _currentNoteIndex + 1,
      'total_notes': _totalNotes,
      'progress_percent': progressPercent,
      'accuracy': _calculateAccuracy(),
    };
  }
  
  /// Reset progress to beginning of song
  void reset() {
    _currentNoteIndex = 0;
    _correctDirections = 0;
    _incorrectDirections = 0;
    _lastDirection = '';
  }
  
  /// Get a description of the current position in the song
  String _getPositionInSong() {
    final noteIndex = _currentNoteIndex % _totalNotes;
    
    if (noteIndex == 0) {
      return "Starting 'Twinkle Twinkle Little Star'";
    } else if (noteIndex == 7) {
      return "At 'How I wonder what you are'";
    } else if (noteIndex == 14) {
      return "At 'Up above the world so high'";
    } else if (noteIndex == 21) {
      return "At 'Like a diamond in the sky'";
    } else if (noteIndex == 28) {
      return "Back to 'Twinkle Twinkle Little Star'";
    } else if (noteIndex == 35) {
      return "Finishing with 'How I wonder what you are'";
    } else {
      return "Note ${noteIndex + 1} of $_totalNotes";
    }
  }
  
  /// Calculate current accuracy percentage
  double _calculateAccuracy() {
    final total = _correctDirections + _incorrectDirections;
    if (total == 0) return 0.0;
    return _correctDirections / total;
  }
  
  /// Get statistics about the current progress
  Map<String, dynamic> getProgressStats() {
    return {
      'current_note': _currentNoteIndex + 1,
      'total_notes': _totalNotes,
      'progress_percent': _currentNoteIndex / _totalNotes,
      'correct_directions': _correctDirections,
      'incorrect_directions': _incorrectDirections,
      'accuracy': _calculateAccuracy(),
    };
  }
}

File: utils/rhythm_trainer.py
import time

class RhythmTrainer:
    """
    Class to handle rhythm training for Twinkle Twinkle Little Star pattern
    """
    def __init__(self):
        # Suzuki Book 1 "Twinkle Twinkle Little Star" bowing pattern
        # Alternating down and up bows for A section (16 changes)
        # Repeated for B section (16 more changes)
        self.twinkle_pattern = ["Down bow", "Up bow"] * 16
        self.current_position = 0
        self.last_direction = None
        self.correct_count = 0
        self.incorrect_count = 0
        self.last_update_time = time.time()
        
    def update_progress(self, current_direction):
        """
        Update progress through the song based on detected bow direction
        
        Args:
            current_direction: Current detected bow direction
            
        Returns:
            Tuple of (status message, current progress)
        """
        # Skip if direction not detected or is "Holding"
        if current_direction in ["Not detected", "Calibrating...", "Holding"]:
            return "Waiting for bow movement...", self.current_position
            
        # Check if this is a new direction (not the same as last recorded)
        if current_direction != self.last_direction and self.last_direction is not None:
            # Check if enough time has passed since last update (to avoid rapid changes)
            current_time = time.time()
            if current_time - self.last_update_time < 0.5:  # 500ms minimum between updates
                return self._get_status_message(), self.current_position
                
            # Direction has changed, check if it matches the expected pattern
            expected_direction = self.twinkle_pattern[self.current_position]
            
            if current_direction == expected_direction:
                self.correct_count += 1
                status = f"Correct! {current_direction} detected."
            else:
                self.incorrect_count += 1
                status = f"Oops! Expected {expected_direction}, but detected {current_direction}."
                
            # Move to next position in pattern
            self.current_position = (self.current_position + 1) % len(self.twinkle_pattern)
            self.last_update_time = current_time
        else:
            status = f"Current bow: {current_direction}, next expected: {self.twinkle_pattern[self.current_position]}"
            
        # Update last direction
        self.last_direction = current_direction
        
        return status, self.current_position
    
    def reset_progress(self):
        """
        Reset progress to beginning of song
        """
        self.current_position = 0
        self.last_direction = None
        self.correct_count = 0
        self.incorrect_count = 0
        self.last_update_time = time.time()
        
    def _get_status_message(self):
        """
        Get status message based on current progress
        
        Returns:
            Status message string
        """
        total = self.correct_count + self.incorrect_count
        if total == 0:
            accuracy = 0
        else:
            accuracy = (self.correct_count / total) * 100
            
        next_expected = self.twinkle_pattern[self.current_position]
        
        return f"Progress: {self.current_position}/32 | Next: {next_expected} | Accuracy: {accuracy:.1f}%"
    
    def get_progress_stats(self):
        """
        Get progress statistics
        
        Returns:
            Dictionary of progress statistics
        """
        total = self.correct_count + self.incorrect_count
        if total == 0:
            accuracy = 0
        else:
            accuracy = (self.correct_count / total) * 100
            
        return {
            "position": self.current_position,
            "total_bows": len(self.twinkle_pattern),
            "correct_count": self.correct_count,
            "incorrect_count": self.incorrect_count,
            "accuracy": accuracy
        }
